samples: matter: Fix NUS switching.

ArekBalysNordic opened the pull request:
There should be a mechanism to disable NUS connection and eventing during the commissioning to Matter process.

Added all lock states to NUS service.
It is needed to send the new state after changing to BLE NUS BT characteristic to allow reading it by the controller.

NordicBuilder commented:
The following west manifest projects have been modified in this Pull Request:

| Name | Old Revision | New Revision | Diff |
| ---- | ------------ | ------------ |------|
| matter | https://github.com/nrfconnect/sdk-connectedhomeip/commit/2a8f797f1f4dc2439d6c4def25db7a4698d1f339 (master) | https://github.com/nrfconnect/sdk-connectedhomeip/pull/268 | [nrfconnect/sdk-connectedhomeip#268/files](https://github.com/nrfconnect/sdk-connectedhomeip/pull/268/files) |

*Note: This message is automatically posted and updated by the Manifest GitHub Action.*

NordicBuilder commented:
## Test specification

### [CI/Jenkins/NRF](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-10918/1/)
- Integration Platforms

### [CI/Jenkins/integration](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-10918/1/)
| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-fw-nrfconnect-chip | X | | |

**test-fw-nrfconnect-chip:** added because there was no .github/test-spec.yml in 'matter'

[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-10918/1/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

kkasperczyk-no commented on samples/matter/common/src/bt_nus_service.cpp:
```diff
@@ -60,20 +60,25 @@ bool NUSService::Init(uint8_t priority, uint16_t minInterval, uint16_t maxInterv
 	mAdvertisingRequest.onStarted = [](int rc) {
 		if (rc == 0) {
-			LOG_INF("NUS BLE advertising started");
+			GetNUSService().mIsStarted = true;
```
I think `GetNUSService()` is not needed, as the field should be accessible directly from this method.

ArekBalysNordic replied:
But then I cannot assign in the simple way that lambda to onStarted method.

kkasperczyk-no commented on samples/matter/common/src/bt_nus_service.cpp:
```diff
@@ -85,6 +90,10 @@ void NUSService::StartServer()
 		return;
 	}
 
+	VerifyOrReturn(bt_conn_auth_cb_register(&sConnAuthCallbacks) == 0);
```
Maybe we should change the type returned by `StartServer` to bool or int? So far there wasn't much that could fail in this method, but now we've got some checks moved from `Init` that actually may fail.

kkasperczyk-no commented on samples/matter/lock/README.rst:
```diff
@@ -234,6 +234,10 @@ The PIN code is different depending on the build type:
 See `Testing door lock using Bluetooth LE with Nordic UART Service`_ for more information about how to test this feature.
 
+.. note::
+   In this sample, there is only one Bluetooth connection available simultaneously.
```
```suggestion
   In this sample, there is only one Bluetooth LE connection available simultaneously.
```

kkasperczyk-no commented on samples/matter/lock/README.rst:
```diff
@@ -581,6 +589,9 @@ To test the :ref:`matter_lock_sample_ble_nus` feature, complete the following st
 
 The Bluetooth LE connection between a phone and the DK will be suspended when the commissioning to the Matter network is in progress or there is an active session of SMP DFU.
 
+To read the current door lock state from the device, read RX Bluetooth LE characteristics.
```
```suggestion
To read the current door lock state from the device, read Bluetooth LE RX characteristic.
```

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-10918
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

greg-fer commented on samples/matter/lock/README.rst:
```diff
@@ -234,6 +234,10 @@ The PIN code is different depending on the build type:
 
 See `Testing door lock using Bluetooth LE with Nordic UART Service`_ for more information about how to test this feature.
 
+.. note::
+   This sample supports one Bluetooth LE connection at a time.
+   Matter commissioning, DFU, and NUS over Bluetooth LE have to be run separately.
```
```suggestion
.. note::
   This sample supports one Bluetooth LE connection at a time.
   Matter commissioning, DFU, and NUS over Bluetooth LE must be run separately.
```
All good, but I would place this in line 223, with appropriate empty lines before and after.

greg-fer commented on samples/matter/lock/README.rst:
```diff
@@ -581,6 +588,9 @@ To test the :ref:`matter_lock_sample_ble_nus` feature, complete the following st
 
 The Bluetooth LE connection between a phone and the DK will be suspended when the commissioning to the Matter network is in progress or there is an active session of SMP DFU.
 
+To read the current door lock state from the device, read Bluetooth LE RX characteristic.
```
```suggestion
To read the current door lock state from the device, read the Bluetooth LE RX characteristic.
```

greg-fer commented on samples/matter/lock/README.rst:
```diff
@@ -581,6 +588,9 @@ To test the :ref:`matter_lock_sample_ble_nus` feature, complete the following st
 The Bluetooth LE connection between a phone and the DK will be suspended when the commissioning to the Matter network is in progress or there is an active session of SMP DFU.
 
+To read the current door lock state from the device, read Bluetooth LE RX characteristic.
+The new lock state is updated after changing the state from any of the following source: NUS, buttons, Matter stack.
```
```suggestion
The new lock state is updated after changing the state from any of the following sources: NUS, buttons, Matter stack.
```

markaj-nordic commented on samples/matter/lock/src/app_task.cpp:
```diff
@@ -636,22 +638,28 @@ void AppTask::LockStateChanged(BoltLockManager::State state, BoltLockManager::Op
 		sLockLED.Set(true);
 #ifdef CONFIG_CHIP_NUS
-		GetNUSService().SendData("Locked", sizeof("Locked"));
+		GetNUSService().SendData("locked", sizeof("locked"));
```
BTW. NUS API would be more concise if the user didn't have to provide the size of a command. Not sure if we can assume null terminated string each time though.

markaj-nordic commented on samples/matter/common/src/bt_nus_service.cpp:
```diff
@@ -60,41 +60,50 @@ bool NUSService::Init(uint8_t priority, uint16_t minInterval, uint16_t maxInterv
-	mAdvertisingRequest.onStopped = []() { LOG_DBG("NUS BLE advertising stopped"); };
+	mAdvertisingRequest.onStopped = []() {
+		bt_conn_auth_info_cb_unregister(&sConnAuthInfoCallbacks);
+		bt_conn_auth_cb_register(NULL);
```
nit: NULL -> nullptr?

markaj-nordic commented on samples/matter/common/src/bt_nus_service.cpp:
```diff
@@ -60,41 +60,50 @@ bool NUSService::Init(uint8_t priority, uint16_t minInterval, uint16_t maxInterv
+		return false;
 	}
 
+	VerifyOrReturnError(bt_conn_auth_cb_register(&sConnAuthCallbacks) == 0, false);
```
nit: is there any difference between `VerifyOrReturnError` and `VerifyOrReturnValue`? Isn't the first one dedicated to CHIP_ERROR type?

markaj-nordic reviewed (approved):
My general thought is that we should not mix the NCS application specific code with Matter SDK code, i.e. NUS Service shouldn't know anything about the Matter. Looks like here we use chip namespaces only to synchronize the chip stack and leverage the VerifyOrReturn...() utilities. To remove these dependencies probably the BLE Arbiter would need to be slightly refactored to provide thread safety, but this is obviously not the scope of this PR.

ArekBalysNordic commented:
@markaj-nordic I addressed your comments and removed most of the chip-related code from NUS. I need to leave BLEArbiter there until it is not separated from the Matter stack. 

After discussion, we decided to disable Bluetooth bonding for the NUS feature. PR has been rebased.
